mp_radix_size_overestimate.c

```diff
@@ -0,0 +1,90 @@
+#include "tommath_private.h"
+#ifdef MP_RADIX_SIZE_OVERESTIMATE_C
+/* LibTomMath, multiple-precision integer library -- Tom St Denis */
+/* SPDX-License-Identifier: Unlicense */
+
+/*
+   Overestimate the size needed for the bigint to string conversion by a very small amount.
+   The error is about 10^-8; it will overestimate the result by at most 11 elements for
+   a number of the size 2^(2^31)-1 which is currently the largest possible in this library.
+   Some short tests gave no results larger than 5 (plus 2 for sign and EOS).
+ */
+
+/*
+   Table of {0, INT(log_2([1..64])*2^p)+1 } where p is the scale
+   factor defined in LTM_RADIX_SIZE_SCALE and INT() extracts the integer part (truncating).
+   Good for 32 bit "int". Set LTM_RADIX_SIZE_SCALE = 61 and recompute values
+   for 64 bit "int".
+ */
+/* *INDENT-OFF* */
+#define LTM_RADIX_SIZE_SCALE 29
+static const uint32_t logbases[65] = {
+           0u,         0u, 0x20000001u, 0x14309399u, 0x10000001u,
+   0xdc81a35u, 0xc611924u, 0xb660c9eu,   0xaaaaaabu,  0xa1849cdu,
+   0x9a209a9u, 0x94004e1u, 0x8ed19c2u,   0x8a5ca7du,  0x867a000u,
+   0x830cee3u, 0x8000001u, 0x7d42d60u,   0x7ac8b32u,  0x7887847u,
+   0x7677349u, 0x749131fu, 0x72d0163u,   0x712f657u,  0x6fab5dbu,
+   0x6e40d1bu, 0x6ced0d0u, 0x6badbdeu,   0x6a80e3bu,  0x6964c19u,
+   0x6857d31u, 0x6758c38u, 0x6666667u,   0x657fb21u,  0x64a3b9fu,
+   0x63d1ab4u, 0x6308c92u, 0x624869eu,   0x618ff47u,  0x60dedeau,
+   0x6034ab0u, 0x5f90e7bu, 0x5ef32cbu,   0x5e5b1b2u,  0x5dc85c3u,
+   0x5d3aa02u, 0x5cb19d9u, 0x5c2d10fu,   0x5bacbbfu,  0x5b3064fu,
+   0x5ab7d68u, 0x5a42df0u, 0x59d1506u,   0x5962ffeu,  0x58f7c57u,
+   0x588f7bcu, 0x582a000u, 0x57c7319u,   0x5766f1du,  0x5709243u,
+   0x56adad9u, 0x565474du, 0x55fd61fu,   0x55a85e8u,  0x5555556u
+};
+/* *INDENT-ON* */
+
+mp_err mp_radix_size_overestimate(const mp_int *a, const int radix, size_t *size)
+{
+   int bit_count;
+   mp_int bi_bit_count, bi_k;
+   mp_err err = MP_OKAY;
+
+   *size = 0u;
+
+   if ((radix < 2) || (radix > 64)) {
+      return MP_VAL;
+   }
+
+   if (mp_iszero(a)) {
+      *size = 2U;
+      return MP_OKAY;
+   }
+
+   if (MP_HAS(S_MP_LOG_2EXPT) && MP_IS_2EXPT((mp_digit)radix)) {
+      /* floor(log_{2^n}(a)) + 1 + EOS + sign */
+      *size = (size_t)(s_mp_log_2expt(a, (mp_digit)radix));
+      /* Would overflow with base 2 otherwise */
+      if (*size > (INT_MAX - 4)) {
+         return MP_VAL;
+      }
+      *size += 1u + 1u + (mp_isneg(a) ? 1u : 0u);
```

Comment:
we discussed this before and @sjaeckel agreed that we should simply write +3

Reply:
Now that you mentioned it I faintly remember, yes. ;-)
Why did I do&hellip;uh, wanted to add a kind of refinement to get an exact value out of it, too. It would need only one big exponentiation smaller than the input, only one single digit multiplication in the normal case and additionaly one very small multiplication with very big inputs. All in all a bit faster and smaller than using `mp_log` but you still have one big exponentiation and if you want to use `mp_log` elsewhere you have that balast on top, so I dropped that idea.

I still have the concept in Pari/GP here:

```text
log2(x)=log(x)/log(2);
numdigits_under_p(a, b) = {
   local(k, k2, n, bn, lbn, la, ldiff, i=0);
   \\ Result "k" fits into a 32-bit integer
   k =  floor(2^29/log2(b)) + 1;
   \\ la = floor(log(a)/log(2)) + 1;
   la = logint(a,2);
   \\ result can and will overflow a 32-bit integer!
   n = ((la *  k) >> 29) + 1;
   \\print("n = " n);
   \\ This is the place where a mp_radix_size_overestimate would return.
   \\ Otherwise: go on.
   \\return (n);
   \\ At this point n >= (floor(log(a)/log(b)) + m) with m > 0
   \\ so we decrement n by one.
   n = n - 1;
   \\ Now bn <= a in almost all cases because
   \\ "m" is not larger than 1 for numbers smaller than ~10^8 bits.
   bn = b^n;

   \\ comparing a, bn can be done by comparing flog2(a), flog2(bn)
   \\ instead which might be computationally cheaper.

   \\ Check the size of "bn" in bits and compute the difference to "a"
   \\ to get the correction necessary
   lbn = logint(bn,2);
   \\ first negatives at "random(2^(2^27))" at b=50,51
   \\ Some empirical tests, although with a small N but with very large
   \\ numbers (<2^(2^31)) gave ldiff = -16 bits as a max. (5 decimal)
   ldiff = la - lbn;

   if(ldiff != 0,
      \\if ( ldiff < 0 , print("ldiff = " ldiff));
      \\ k2 can be zero if ldiff <= round(log_2(b))
      k2 = (k * ldiff) >> 29;
      \\print("k2 = " k2 );
      if( k2 > 0,
         bn = bn * (b^k2);
      );
      if( k2 < 0,
         \\ exact division.
         bn = bn / (b^(-k2));
      );
      n = n + k2;
   );

   if(a == bn,
      return(n + 1);
   );
   if(a > bn,
      while( a > bn,
        bn = bn * b;
        n = n + 1;
      );
      \\ a < bn, return n
      return(n);
   );
   if(a < bn,
      while( a < bn,
        \\ again, exact division
        bn = bn / b;
        n = n - 1;
      );
   );

   \\ a >= bn, return n + 1.
   return(n + 1);
}
```
It's a rough sketch but probably better than my english description ;-)

Reply:
Hmm I am not sure. If we could get a fast and small exact version I would prefer that since we could simply replace mp_radix_size. So we would only get mp_expt_n additionally? Maybe that would be ok, I think mp_expt is pretty small.

Reply:
@czurnieden how would it compare to the currently used mp_radix_size? This is already faster than the division approach we did before. But not particularly small. Maybe it makes sense to offer the very downstripped mp_radix_size_overestimate as proposed here since it is really only needed for allocation size estimation.

Reply:
> So we would only get mp_expt_n additionally?

Yes. Instead of both `mp_log` and `mp_expt`

> how would it compare to the currently used mp_radix_size?

Not implemented in LTM, can't say. But there is negligible overhead beside the single big exponentiation. But there is not much use beside `mp_radix_size`.

> Maybe it makes sense to offer the very downstripped mp_radix_size_overestimate as proposed here since it is really only needed for allocation size estimation.

So you want to remove the exact `mp_radix_size`, which is basically just a convenient wrapper for `mp_log` now and replace it with a note in the documentation to use `mp_log` directly if an exact result is needed or `mp_radix_size_overestimate` for all of the other 99.999something% involving number conversions?
I'm all for it.

Reply:
I don't want to remove mp_radix_size. We discussed this before. The question is only if we add the overestimating function or if we change the internals of mp_radix_size instead.